### ISSUE TITLE
https login causes cockpit.service to crash: src/tls/connection.c:318: buffer_write_to_tls: Assertion `buffer_valid (self)' failed.

### ISSUE THREAD

**Issue by martinpitt** [cockpit-project/cockpit#12987]
Since PR #12902 , a manual login with Firefox with https pretty much immediately crashes cockpit.socket after activation:
```
cockpit-tls: gnutls_handshake failed: A TLS fatal alert has been received.
cockpit-tls: src/tls/connection.c:318: buffer_write_to_tls: Assertion `buffer_valid (self)' failed.
cockpit.service: Main process exited, code=killed, status=6/ABRT
```

This blocks our release today.

**Comment by mvollmer** [comment]
For me, the crash does not always happen.  Sometimes Cockpit loads half-way and then Firefox freezes with 100% CPU.  Killing cockpit-tls allows Firefox to recover.

**Comment by martinpitt** [comment]
I stared at this for a while, and I can't spot the bug yet. 
```
src/tls/connection.c: buffer_read_from_tls begin (client-to-ws/0x2f48/0x300c, 0xcb4ec0)
src/tls/connection.c:   get_iovecs (0x7fae127f3c00, 1, 0x7fae127f3c90, 0x300c, 0x6f48)
src/tls/connection.c:     iov[0] = { 0x300c, 0xff4 };  start = 0x4000;
src/tls/connection.c:     return 1;
src/tls/connection.c: buffer_read_from_tls end (client-to-ws/0x2f48/0x307a, 0xcb4ec0)
src/tls/connection.c: buffer_write_to_tls begin (ws-to-client/0x53d3/0x93d3, 0xcb4ec0)
src/tls/connection.c:   get_iovecs (0x7fae127f3c00, 1, 0x7fae127f7ca8, 0x53d3, 0x93d3)
src/tls/connection.c:     iov[0] = { 0x13d3, 0x2c2d };  start = 0x8000;
src/tls/connection.c:     return 1;
src/tls/connection.c: buffer_write_to_tls end (ws-to-client/0x53c9/0x93d3, 0xcb4ec0)
cockpit-tls: src/tls/connection.c:307: buffer_write_to_tls: Assertion `buffer_valid (self)' failed.
```
The assertion is clearly justified, as 0x93d3 - 0x53c9 == 16394 which exceeds the buffer. So something in the ring buffer/iovec house keeping is buggy. This has taken me too long already for an "under the gun" fix, so for now I'll just revert #12902 and we postpone that to the next release.

**Comment by martinpitt** [comment]
Untagging release-blocker, we'll go with the revert for now in PR #12988.
